### PR TITLE
ci: move semver-checks to its own job

### LINKED ISSUE #2968: CI is failing on `master` branch (semver-checks)
CI is passing for PRs but failing on `master` because it's failing to find a base branch.

I believe it needs to be pulled out of `.github/workflows/rust.yml` into its own file which has `on: pull_request:` but no `on: branches:`. But this is inherently difficult/annoying to test.

### PR DESCRIPTION
**Pull request by storopoli**
This is necessary because it needs to run only on pull_request, and not on push events.

Closes #2968.

### CODE REVIEW

File: .github/workflows/semver-checks.yml
Comment (Kixunil) [inline .github/workflows/semver-checks.yml:2]: I believe you need to remove this.

Comment (storopoli) [reply .github/workflows/semver-checks.yml:2]: Yes, done.

File: .github/workflows/semver-checks.yml
Comment (Kixunil) [inline .github/workflows/semver-checks.yml:32]: I suggest that rather than having two jobs just to fill this we should just call `rustup default $(cat nightly-version)` here instead of the external action

**Review by Kixunil — DISMISSED** [review]
ACK cc4de3b213c12f20d8f04d0f29b018dc8326b50e

File: .github/workflows/semver-checks.yml
Comment (apoelstra) [reply .github/workflows/semver-checks.yml:32]: Oh, nice. If that works we should definitely do it.

Comment (apoelstra) [reply .github/workflows/semver-checks.yml:32]: We could maybe fold this into `test.sh` if we added some sort of warning and found a way to put it back when we were done.

Comment (storopoli) [reply .github/workflows/semver-checks.yml]: Ok, good suggestion. I've reduced to one job in f8c91b0.
Still, I don't know how to do it in one step since the `with:` yml part cannot take the `cat` command. It is a 2-step thing:

1. install `rustup` with `dtolnay/toolchain`
2. set the `rustup default` with the output of `cat`

Comment (storopoli) [reply .github/workflows/semver-checks.yml]: > We could maybe fold this into `test.sh` if we added some sort of warning and found a way to put it back when we were done.

You mean add it to https://github.com/rust-bitcoin/rust-bitcoin-maintainer-tools/blob/master/ci/run_task.sh ?

**Review by apoelstra — APPROVED** [review]
ACK f8c91b067b4851724fe68b3c7a41ec0181d04853

**Comment by tcharding** [comment]
> This is necessary because it needs to run only on pull_request, and not on push events.

Why is this the claim please? If a PR is made that has no api changes then we get no message, a later force push then does a bunch of api changes, I would have thought  we then want the warning message?

**Comment by storopoli** [comment]
> > This is necessary because it needs to run only on pull_request, and not on push events.
> 
> Why is this the claim please? If a PR is made that has no api changes then we get no message, a later force push then does a bunch of api changes, I would have thought we then want the warning message?

From the issue #2968:

> CI is passing for PRs but failing on master because it's failing to find a base branch.

https://github.com/rust-bitcoin/rust-bitcoin/issues/2968#issue-2392715676

This is because of https://github.com/rust-bitcoin/rust-bitcoin/blob/1c96da080ddf8fbeb742398bf835d1815ef00721/contrib/check-semver.sh#L22-L29

**Comment by apoelstra** [comment]
>Why is this the claim please? If a PR is made that has no api changes then we get no message, a later force push then does a bunch of api changes, I would have thought we then want the warning message?

The issue is that it's triggering on pushes to master.

The `push` target isn't what gets run on force-pushes (except direct pushes to master I guess? which should never happen). The `pull_request` target is. Or maybe both targets get run in this case... I remember some "CI is running twice" optimization from many years ago related to this.

**Comment by tcharding** [comment]
> The push target isn't what gets run on force-pushes ... The pull_request target is

My bad, I was misunderstanding how those targets worked. Thanks, I rescind my concerns.

**Review by tcharding — APPROVED** [review]
ACK f8c91b067b4851724fe68b3c7a41ec0181d04853

**Comment by tcharding** [comment]
I see that we are missing docs from the `workflows/README.md` file for the original API job, which if it existed should have been updated as part of this PR. Since its missing we can do it as a follow up - I created #2975.

**Review by Kixunil — APPROVED** [review]
ACK f8c91b067b4851724fe68b3c7a41ec0181d04853

**Comment by storopoli** [comment]
> I see that we are missing docs from the `workflows/README.md` file for the original API job, which if it existed should have been updated as part of this PR. Since its missing we can do it as a follow up - I created #2975.

Yeah let's merge this since it has 3 ACKs ASAP so that we fix CI on master.
After that we'll update the README.

File: .github/workflows/semver-checks.yml
Comment (Kixunil) [inline .github/workflows/semver-checks.yml:18]: Oh, wait, this unpins the nightly. But let's merge this to unblock the CI and fix it afterwards.

Comment (tcharding) [reply .github/workflows/semver-checks.yml:18]: Something is odd here, we install the nightly toolchin but use the version in `nightly-version` in `contrib/check-semver.sh`. I'm surprised the script runs - unless we just happen to have the same nightly version as `dtonlnay`'s tool installs which I doubt?

Comment (tcharding) [reply .github/workflows/semver-checks.yml:18]: FTR still +1 from me to merge and fix later.

Comment (Kixunil) [reply .github/workflows/semver-checks.yml:18]: Actually it's pinned in the script. This action is a bit misleading, we only need rustup on the machine.

Comment (storopoli) [reply .github/workflows/semver-checks.yml:18]: Yes, we can put anything in here. It's just to have rustup setup so that the next step does not fail.

Comment (storopoli) [reply .github/workflows/semver-checks.yml:18]: These steps

https://github.com/rust-bitcoin/rust-bitcoin/blob/f8c91b067b4851724fe68b3c7a41ec0181d04853/.github/workflows/semver-checks.yml#L17-L21

Comment (tcharding) [reply .github/workflows/semver-checks.yml:18]: Oh woops, missed that bit. Thanks

**Pull request merged**
